public/app/features/dashboard/state/PanelQueryState.ts

```diff
@@ -0,0 +1,168 @@
+import {
+  DataSourceApi,
+  DataQueryRequest,
+  PanelData,
+  LoadingState,
+  toLegacyResponseData,
+  isSeriesData,
+  toSeriesData,
+  DataQueryError,
+} from '@grafana/ui';
+import { getProcessedSeriesData } from './PanelQueryRunner';
+import { getBackendSrv } from 'app/core/services/backend_srv';
+import isEqual from 'lodash/isEqual';
+
+export class PanelQueryState {
+  // The current/last running request
+  request = {
+    startTime: 0,
+    endTime: 1000, // Somethign not zero
+  } as DataQueryRequest;
+
+  // The best known state of data
+  data = {
+    state: LoadingState.NotStarted,
+    series: [],
+  } as PanelData;
+
+  sendSeries = false;
+  sendLegacy = false;
+
+  // A promise for the running query
+  private executor: Promise<PanelData> = {} as any;
+  private rejector = (reason?: any) => {};
+  private datasource: DataSourceApi = {} as any;
+
+  isRunning() {
+    return this.data.state === LoadingState.Loading; //
+  }
+
+  isSameQuery(ds: DataSourceApi, req: DataQueryRequest) {
+    if (this.datasource !== this.datasource) {
+      return false;
+    }
+
+    // For now just check that the targets look the same
+    return isEqual(this.request.targets, req.targets);
+  }
+
+  getCurrentExecutor() {
+    return this.executor;
+  }
+
+  cancel(reason: string) {
+    const { request } = this;
+    if (!request.endTime) {
+      this.rejector('Query Canceled:' + reason);
+      request.endTime = Date.now();
+    }
+
+    // Cancel any open HTTP request with the same ID
+    if (request.requestId) {
+      getBackendSrv().resolveCancelerIfExists(request.requestId);
+    }
+  }
+
+  execute(ds: DataSourceApi, req: DataQueryRequest): Promise<PanelData> {
+    this.request = req;
+
+    // Return early if there are no queries to run
+    if (!req.targets.length) {
+      console.log('No queries, so return early');
+      this.request.endTime = Date.now();
+      return Promise.resolve(
+        (this.data = {
+          state: LoadingState.Done,
+          series: [], // Clear the data
+          legacy: [],
+          request: req,
+        })
+      );
+    }
+
+    // Set the loading state immediatly
+    this.data.state = LoadingState.Loading;
+    return (this.executor = new Promise<PanelData>((resolve, reject) => {
+      this.rejector = reject;
```

Comment:
Main goal is to hold on to the promise (this.executor) so it can be returned for duplicate queries.

Is it safe to grab this and call reject in cancel()?  

Is there a better modern async way to do this?

Reply:
think this is ok